File: threedollar-api/src/main/java/com/depromeet/threedollar/api/controller/ApiResponse.java
package com.depromeet.threedollar.api.controller;

import com.depromeet.threedollar.domain.exception.ErrorCode;
import lombok.*;

@ToString
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class ApiResponse<T> {

	public static final ApiResponse<String> SUCCESS = success("OK");

	private String resultCode;

	private String message;

	private T data;

	public static <T> ApiResponse<T> success(T data) {
		return new ApiResponse<>("", "", data);
	}

	public static <T> ApiResponse<T> error(ErrorCode errorCode) {
		return new ApiResponse<>(errorCode.getCode(), errorCode.getMessage(), null);
	}

	public static <T> ApiResponse<T> error(ErrorCode errorCode, String message) {
		return new ApiResponse<>(errorCode.getCode(), message, null);
	}

}

File: threedollar-admin/src/main/kotlin/com/depromeet/threedollar/admin/common/dto/ApiResponse.kt
package com.depromeet.threedollar.admin.common.dto

import com.depromeet.threedollar.common.exception.ErrorCode

data class ApiResponse<T>(
    val resultCode: String,
    val message: String,
    val data: T?
) {

    companion object {
        val OK = success("OK")

        fun <T> success(data: T): ApiResponse<T> {
            return ApiResponse("", "", data)
        }

        fun error(errorCode: ErrorCode): ApiResponse<Nothing> {
            return ApiResponse(errorCode.code, errorCode.message, null)
        }

        fun error(errorCode: ErrorCode, message: String?): ApiResponse<Nothing> {
            return ApiResponse(errorCode.code, message ?: errorCode.message, null)
        }
    }

}
